### ISSUE TITLE
UIPageViewController

### ISSUE THREAD

**Issue by jcampbell05** [Simbul/baker#529]
I have got a simple implementation working, I still have to fix up a few other things but hopefully with the new swiping page thing this controller offers in iOS 6 this will make behaviour of going between pages easier.

I was thinking of introducing a new property to the manifest that dictates the page turning functionality. I think it should be something like "-baker-page-turn-style" with the options being: "Swipe", "Page" and "ScrollSwipe".  The Swipe mode will use the new page mode of the pageviewcontroller, only avlaible fo iOS 6 + (i.e like the home screen on iOS), The Page mode will be like iBooks and the ScrollSwipe will be how baker implements it now. The default to maintain backwards compatibility will be ScrollSwipe.

If a swipe mode is chosen that the current iOS doesn't support then the app will drop to ScrollSwipe. (i.e page turn mode will work on iOS 5 and above so iOS 4 devices and below will use ScrollSwipe)

An image on the progress so far:

![Screenshot](http://i48.tinypic.com/35almde.png)

your thoughts?

**Comment by folletto** [comment]
That's a great work! Where are you implementing it right now? `master` or `shelf`? :)

Could you also detail a little bit more the difference between "Swipe" and "Page"? I'm not entirely sure I got it. :)
For the rest, I think that the JSON label you choose is ok, I would probably give different names to the properties but I'll wait to understand better the difference between the two. :)

Good job! :)

**Comment by jcampbell05** [comment]
Master at the moment haven't quite got my head round github enough to be able to edit paticular branches :) if its an issue I can always send you the code.

Okay basically in an essence:

ScrollSwipe - Will be how we implement it in Baker today, with the scrollview wrapping a bunch of webviews
Page - Will be the new UIPageViewController in Page Turn Mode which will look like iBooks
Swipe - Will be the new UIPageViewController in Scroll Mode which will be like how baker does it today but using the native functionality of the UIPageViewController to achieve it.

Maybe we could come up with better names for ScrollSwipe and Swipe. I was thinking originally Legacy or Baker, Page and Swipe or Scroll but wasn't sure.

One question where is the part of the code where the scrollview gets it's first webview when the first page is loaded, at the moment I can only get pages to appear in the page view if I make a fake page and then load in the current webview for the current page by doing the page turn gesture. Also how does the page turning work currently ? I've got a vague understanding :) of how the code works through trial and error :)

James

**Comment by folletto** [comment]
Right now Baker uses a `UIScrollView`, with _pagination_, and places the `UIWebViews` over it. Everything is then managed automatically from an interaction perspective, we "just" have to instantiate the `UIWebView` accordingly (using one of the two modes, screenshots or three-cards). A lot of the code there handles the smooth transition when the user swipes the scrollview.

About the modes, if `swipe` is similar to the native functionality, what are the pros of using it instead of the native one? :)

**Comment by jcampbell05** [comment]
Ah I see, I'm right now trying to break out the instanation of the Webviews, so that I can get the correct one for the next page :) Which is the biggest headache. I also need to make the pagecontroller detect when the first webview is created so it can display it as the first page.

The only pros to keeping the current implementation is backwards compatibility with older Devices. At the moment the Native one only works with iOS 6 above in Scroll Mode :) and iOS 5 above in page turn mode. We can't fully move over to it unless we want to cut off people below iOS 5. (And disable the use of the page scrolling for people below iOS 6)

**Comment by jcampbell05** [comment]
I think I have figured out the UIWebView code, correct me if i'm wrong. 

In three-cards:

When you turn the page, the scroll view puts in the boilerplate loading UI and requests the next page from the book.

The code checks if the user hasn't moved so much that the pages it has preloaded are invalid (i.e they move 2 pages instead of 1) and that the next page has loaded.

If they are loaded then the pointers are reshuffled so the next page pointer contains null and current page contains what was originally in the next page pointer.

 the webview is injected into the scrollview and the pointers and shown as the current view. If not then the pointer reshuffling occurs still but the next page is placed onto a queue and loaded.

 Once loaded the webview is assigned to the current page pointer and injected into the view.

If we arn't using three card mode then we simply inject the screenshot image if available until the webview is loaded and we can inject that. We then take a screen shot if not.

**Comment by folletto** [comment]
Yes, seems correct to me. :)

**Comment by jcampbell05** [comment]
Awesome I think that flow could somehow be hacked into UIPageViewController :)

**Comment by jcampbell05** [comment]
Okay made some progress it works now :) Just need to get screenshots working. Then I'll probarly have to restructure the code :) so that it properly integrates with the three-cards system as well. I'll also implement some navigation methods for the UIPageView as it can't use the scroll views.

 We will probably have to abstract this, I was thinking that basically the UIScrollView and UIPageViewController could be treated like wrappers and implement the required Protocol which then the main BakerViewController will use to talk to them and give them the WebViews once they are loaded and also inform them of any other state changes like the app launching, closing etc etc. The Wrappers could then request the next page etc as well :).

Then basically the app reads the manifest and works out which Wrapper it needs to load (UIPageViewController or UIScrollView) based on the type of page turning.

I think we will end up with a lot of ifs and elses otherwise.

The Scroll mode on UIPageViewController works but is a little buggy at the moment.

![Screenshot](http://i47.tinypic.com/2njal1e.png)

**Comment by folletto** [comment]
Impressive work. :)

May I nudge you a bit to use GitHub to show the code to us (and the future Pull Request)? It's really little effort and simplify things a whole lot (and GitHub support is great if you need info). :)

**Comment by jcampbell05** [comment]
Awesome, would it be okay if I put this into another branch for now, as this could break the code for the other branches ? I will upload what I've got later today :) to show it working

**Comment by folletto** [comment]
Yes, a branch would be ideal (it will also help you to manage changes when you merge back into `shelf` or `master`) and then the pull request will help us a big deal. :)

But, no hurry. :)

**Comment by jcampbell05** [comment]
Okay I'll create a branch "PageController" and submit it so you'll at least have what I've done so far. And I'll continue with abstracting the View Flow and then once stable enough the pull request :)

**Comment by jcampbell05** [comment]
Okay I've created a new branch on my fork and I am ready to create a new pull request for what I have got so far (Be warned it is very messy and hacky). Could you create a branch on your side called "page-view-controller". As I don't really want to put it in any of the other branches, and github does not let you do branch requests.

James

**Comment by folletto** [comment]
It should be fine creating your branch and linking it here for now, because in that way we can see the progress and test it. Once ready you merge with master and so it will be a pull request to master with your branch merged into it. ;)

**Comment by jcampbell05** [comment]
Okay sounds like the best way to do it :)

Anyways the branch as it is at the moment is here
https://github.com/jcampbell05/baker/tree/page-view-controller

Still got a lot of work to do :) but this is the hacky implementation until I have finished abstracting the webview wrapper :)

I've been thinking of introducing a hybrid mode. So basically it would preload the previous and next page like three cards does, but in addition to that would also show a screenshot of the page just before it's loaded like screenshot mode. So if the user navigated to a page that hadn't been or had not finished preloading they would get a nice screenshot this would especially look nice with the page view controller :). 

James

**Comment by Simbul** [comment]
First of all, kudos for the work! I haven't looked at the code yet but that's definitely a major feature you're introducing there :)

My suggestion would be to keep the pull request changes at a minimum (i.e. keeping the hybrid mode out of it).
This would make it easier for you to get to the point where you can submit a proper pull request and it would make it easier for us to review the pull request and bring it in.

**Comment by jcampbell05** [comment]
Ahh, Okay :) I will do that, sorry I get a bit carried away :) when things start getting awesome :). I'll just finish off my abstraction of the view so that the old scrollview and new page view are interchangeable, which should make it less buggy and more manageable.

**Comment by jcampbell05** [comment]
Just a small progress update :) reworked a lot of your view based code. It seems to be a lot more quick and nifty still got to implement the wrappers that let the user swipe between pages. I've almost got the page stuff it's self working, just gotta do screenshots and the webview stuff :)

**Comment by jcampbell05** [comment]
I've noticed that the book title is incorrect in the manifest. 

Anyways I've got my page controller working :) I am reimplementing stuff for the new View System. I have left is link handling, anchors, scroll direction remembering, better orientation, old baker navigation system, index view, gestures and screenshots.

**Comment by folletto** [comment]
Yes, it doesn't match. Whops. :)

Thanks for the updates, you're making us following you very simpler in this way. ;)

**Comment by jcampbell05** [comment]
:) oops.

It's okay :) I've got iOS 4.0 compatibility working (So if a book uses the page turn transition on iOS4 it will just drop back to using the Scroll View) and it now calculates page size correctly.

:) Just got gestures and index view. Anchors, Scroll Offset, Screenshots and Link Handling Left. A lot of it is in the same cluster of issues really :) I'll probably submit a pull request once all of this is implemented so you can take a look and refine and optimise it all (it can lag occasionally)

**Comment by folletto** [comment]
Ok! :D

**Comment by jcampbell05** [comment]
Okay I've got these left to for the new System
- Anchor Management
- Last Scroll Index Remembering
- Index View Show / Hide
- Fix crashes to do with the links to other pages in the same book and the UIPageViewController Wrapper
- Gestures (Mainly for index view and scroll view)
- Scroll View
- Major Bug Fixing

The Original Scroll View is very buggy and not fully implemented due to the new way I am doing it I had to break it quite heavily. Also the screenshots are buggy especially with longer pages. The Page sizes aren't calculated properly yet either.

I've noticed that with the impending Baker 4.0 release you have now pushed the BakerShelf to the master branch. 

This fork is based on the old 3.X Branch and thus doesn't have the nice things like the classes to manage the state of the book like the height rather than polluting the NSUserDefaults and the customisable modal page button. As well as being a lot more clean :).

Obviously I would like to do more work on this but I think moving forward it would be best to maybe issue a pull request for you guys so you can put it in its own branch ? This way it can be kept in sync with the latest Baker more easily (So Its easier to incorporate changes such as that), I would happily keep on developing it with you guys :)

Is that okay with you ?

**Comment by folletto** [comment]
We don't merge "breaking" features on master until they are ready, but we can use that to keep track of development. :)

**Comment by jcampbell05** [comment]
Yes :) Hence why in my fork I have it as a separate branch so I can sync my master branch with yours :)

Awesome :)  Well I just need to update my code to Baker 4 and fix the bugs etc :)

**Comment by jcampbell05** [comment]
Okay I've restarted this, to base it on the Baker 4 branch :) as so much had changed it was better to start from scratch. 

Based on my experimentation the following needs to happen in the Code Base for us to be able to support using a UIPageViewController.
- We need to abstract the Navigation Controller, So by default the one with baker will be the ScrollView. It should have a similar set of methods and delegate to that of the UPageNavigationController.
- The Actual Web views and Loading Pages need to be moved into a Special Separate View Controller rather than being created automagically by functions :) Basically the Abstracted Navigation Controllers to display above will be handed a Page View Controller which will then show the loading screens and fade in the web view. The Scroll view will show them next to each other and let you scroll through them and the UIPageController will let you turn it like a page (Or swipe if setup correctly on iOS 6 and above). 
- We need to add a configuration setting into the manifest to switch between the two styles of navigation.
- Link Handling, Gestures and Screenshot rendering need to be reworked for the abstracted Navigation view and the new specially created Page View Controller :) 

I have the old tweaked code around somewhere for examples :) Should I give you the above changes in bits and pieces so we can slowly move towards implementing the new page turn effect, I think this would be a better way forward and would prevent me needing to rewrite things :) as much

Moving to ARC would help this along but isn't necessary.

**Comment by folletto** [comment]
Yep, step by step is definitely an excellent approach. :)

**Comment by jcampbell05** [comment]
Cool, I'll start off with the smaller bits :)
